### PR TITLE
Adaptive node protocol

### PR DESCRIPTION
**Pull request by nolash**
obsoletes https://github.com/ethersphere/SWIPs/pull/11

### CODE REVIEW

**Comment by nagydani** [comment]
@nolash , I think, the latest commit, namely
https://github.com/ethersphere/SWIPs/pull/12/commits/1aafb847aa2758be2ef8a976587da81fa8f8fbbd
is unnecessary. It is contingent on your proposed changes to SWIPs in general, which I think are somewhat controversial. Also, I am wondering why you decided to use a different email address?
Would you be willing to revert that last commit?

**Comment by nolash** [comment]
@nagydani I'm not sure what you mean; the latest commit is https://github.com/ethersphere/SWIPs/pull/12/commits/bd568752403056c1097b0fdafb738fc31350f38e

I don't use that email address, it's only for evil google aliasing.

**Comment by nagydani** [comment]
Okay, then it's not the latest, but it is the commit that I linked, which makes it compliant with the changes in the template that are still pending.

File: SWIPs/swip-0.md [hunk 1]
```diff
@@ -8,9 +8,9 @@ created: 2019-07-08
 updated: 2019-07-09
```
Comment (decanus) [inline SWIPs/swip-0.md]: This seems like it should be in another PR, can we move it please?

File: SWIPs/swip-adaptive-node-protocol.md [hunk 2]
```diff
@@ -0,0 +1,131 @@
...
+Capabilities {
+	Caps []Capability
+}
+```
+
+_Note that the RLP serialization of this structure in the golang implementation adds an extra list entry for every struct encapsulation. This "list of capabilities" will in reality be a "list of list of capabilities," where the former list only has one element._
+
+Furthermore, the bzz handshake message itself should also embed this message instead of the `Light` field that is currently there. Incidentally, in the golang implementation, the `BzzAddr` object transmitted in the bzz handshake is also (part of) the peer record stored by the kademlia peer database. A practical approach is therefore to embed the `Capabilities` information in the same structure. `BzzAddr` thus becomes:
+
+type BzzAddr struct {
```
Comment (diegomasini) [inline SWIPs/swip-adaptive-node-protocol.md]: Missing block code delimiters (```)

File: SWIPs/swip-adaptive-node-protocol.md [hunk 3]
```diff
@@ -0,0 +1,131 @@
...
+	OAddr        []byte
+	UAddr        []byte
+	Capabilities *Capabilities
+}
```
Comment (diegomasini) [inline SWIPs/swip-adaptive-node-protocol.md]: Missing block code delimiters (```)

File: SWIPs/swip-0.md (hunk 1 above)
Comment (oskarth) [reply SWIPs/swip-0.md]: +1

File: SWIPs/swip-adaptive-node-protocol.md [hunk 4]
```diff
@@ -0,0 +1,131 @@
...
+title: Adaptive nodes capabilities protocol
+author: Louis Holbrook <dev@holbrook.no> (https://holbrook.no), Tim Bansemer (editorial) @FantasticoFox <tim@ethswarm.org>
+status: Draft
+type: Track Networking
+created: 2019-07-11
+---
+
+## Abstract
+
+Adaptive node operation means a node should be able to change the capabilities it has according to changes in available resources.
```
Comment (oskarth) [inline SWIPs/swip-adaptive-node-protocol.md]: I'd specify, "available resources, such as bandwidth, connectivity ,...". This helps guide the reader what it's about

File: SWIPs/swip-adaptive-node-protocol.md [hunk 5]
```diff
@@ -0,0 +1,131 @@
...
+
+## Motivation
+
+The peers of the node may need to adjust their behavior towards the node when these capabilities change. Therefore, a protocol is required to inform the peers of changes in capabilities.
+
+## Specification
+
+Upon initial connection, capabilities can be transmitted already within the handshake message.
```
Comment (oskarth) [inline SWIPs/swip-adaptive-node-protocol.md]: Perhaps this can point to spec that outline the handshake message?

File: SWIPs/swip-adaptive-node-protocol.md [hunk 6]
```diff
@@ -0,0 +1,131 @@
...
+
+## Abstract
+
+Adaptive node operation means a node should be able to change the capabilities it has according to changes in available resources.
+
+## Motivation
+
+The peers of the node may need to adjust their behavior towards the node when these capabilities change. Therefore, a protocol is required to inform the peers of changes in capabilities.
+
+## Specification
```
Comment (oskarth) [inline SWIPs/swip-adaptive-node-protocol.md]: The first part reads more like an introduction, perhaps it can be under such a header? Then more clear definitions under Definitions header, Flow, Wire/data layout etc

File: SWIPs/swip-adaptive-node-protocol.md [hunk 7]
```diff
@@ -0,0 +1,131 @@
...
+
+Upon initial connection, capabilities can be transmitted already within the handshake message.
+
+Capabilities are combinational. Some combination may make more sense than other. Interpreting which combinations are feasible and which are not should be handled by respective nodes, and not be hardwired in the protocol.
+
+For this reason, a bit vector seems the most logical choice. The bit vector can represent the adaptive node operation state within the node, and be embedded into the protocol message as-is.
```
Comment (oskarth) [inline SWIPs/swip-adaptive-node-protocol.md]: "seems" is a bit a of a weasel word in a spec. Better to just state what it does

File: SWIPs/swip-adaptive-node-protocol.md [hunk 8]
```diff
@@ -0,0 +1,131 @@
...
+Since the capabilities may have to be set from different system modules, the network module should provide an API for registering them. The following structure is proposed:
```
Comment (oskarth) [inline SWIPs/swip-adaptive-node-protocol.md]: why proposed? as opposed to "this is the structure"

File: SWIPs/swip-adaptive-node-protocol.md [hunk 9]
```diff
@@ -0,0 +1,131 @@
...
+
+* the first byte is the identifier of the module (0x00 is bzz, and built in)
```
Comment (oskarth) [inline SWIPs/swip-adaptive-node-protocol.md]: this might be clearer if laid out as ascii data layout or so, then explain further

e.g. see rfc style specs

File: SWIPs/swip-adaptive-node-protocol.md [hunk 10]
```diff
@@ -0,0 +1,131 @@
...
+* the second byte is the length `n` of the bit vector containing the flags
+* the last `n` bytes is the bit vector containing the flags
+
+For the individually capability flags we will initially be choosing from, we can see an example of two distinct byte pairs representing tematic categories:
```
Comment (oskarth) [inline SWIPs/swip-adaptive-node-protocol.md]: individual

this sentence can be more concise and to the point

File: SWIPs/swip-adaptive-node-protocol.md [hunk 11]
```diff
@@ -0,0 +1,131 @@
...
+
+```
+BZZ/SYNC: 			// (registered by default)
```
Comment (oskarth) [inline SWIPs/swip-adaptive-node-protocol.md]: point to relevant specs like bzz/sync and pss (right below this section)

File: SWIPs/swip-adaptive-node-protocol.md [hunk 12]
```diff
@@ -0,0 +1,131 @@
...
+- retrieve		0x00020001
```
Comment (oskarth) [inline SWIPs/swip-adaptive-node-protocol.md]: all these capabilities should be elaborated on right after in terms of what they mean

File: SWIPs/swip-adaptive-node-protocol.md [hunk 13]
```diff
@@ -0,0 +1,131 @@
...
+
+Furthermore, the bzz handshake message itself should also embed this message instead of the `Light` field that is currently there. Incidentally, in the golang implementation, the `BzzAddr` object transmitted in the bzz handshake is also (part of) the peer record stored by the kademlia peer database. A practical approach is therefore to embed the `Capabilities` information in the same structure. `BzzAddr` thus becomes:
+
+type BzzAddr struct {
+	OAddr        []byte
+	UAddr        []byte
+	Capabilities *Capabilities
+}
+
+Consequently the `LightNode` flag is removed from the `HandshakeMsg` message.
```
Comment (oskarth) [inline SWIPs/swip-adaptive-node-protocol.md]: this is hard to parse, "is removed"? the spec should state what is the case, if things are changing or there's a diff from previous versions, I suggest we call that out in a separate section

File: SWIPs/swip-adaptive-node-protocol.md [hunk 14]
```diff
@@ -0,0 +1,131 @@
...
+
+```
+HandshakeMsg {
+	Version		uint64
+	NetworkID	uint64
+	Addr		*BzzAddr
+}
+```
+
+For simplicity, arrays of bools are used to represent the flags in interfacing code. In the JSON-RPC API a `Capability` is represented in the following manner:
```
Comment (oskarth) [inline SWIPs/swip-adaptive-node-protocol.md]: implementation detail, perhaps add to implementation suggestion section to keep spec more concise?

File: SWIPs/swip-adaptive-node-protocol.md [hunk 15]
```diff
@@ -0,0 +1,131 @@
...
+For example, a `Capability` with id 42 and three flags with values `true, false, true` this becomes:
+
+```
+{
+	"Id": 42,
+	"Cap": [true, false, true],
+}
+```
+
+The network module API can merely consist of:
```
Comment (oskarth) [inline SWIPs/swip-adaptive-node-protocol.md]: again, "can" - I want to know what it does
